File: Python/9095_1,2,3더하기_dp.py
if __name__ == '__main__':
	T = int(input())
	for _ in range(T):
		N = int(input())
		
		arr = [0]*(N+1)
		arr[0] = 1 # 초기값 1로 설정
		for i in range(1,4): # +1부터 +3으로 이루어진 합의 수 구할 것
			for n in range(i,N+1): # i부터 채우는 이유는 i보다 적은 값은 숫자 i로 구성할 수 없기 때문
				tmp = 0
				for j in range(1,i+1): # 현재 값은 이전 i개의 배열 값 더해준 값
					tmp += arr[n-j]
				arr[n] = tmp
				
		print(arr[N])

File: C++/9095_1,2,3더하기_dp.cpp
#include <iostream>
using namespace std;

int main() {
	int T;
	cin >> T;
	for(int t = 0; t < T; t++){
		int n;
		cin >> n;
		
		int arr[n+1];
		arr[0] = 1; // arr[0] = 1은 초기값
		
		for(int i = 1; i < 4; i++){ // 1이 들어가는 경우의 수부터 3이 들어가는 경우의 수까지 따지기 위함
			for(int j = i; j < n+1; j++){ // 현재 i보다 작은 값은 i로 만들 수 없기 때문에 i부터 시작해야함
				arr[j] = 0;
				for(int k = 1; k <= i; k++){ // 이전 i개의 배열 값을 더해줌
					arr[j] += arr[j-k];
				}
			}
		}
		
		cout << arr[n] << '\n';
	}
	return 0;
}
